### PR TITLE
Include information on sub-layer parameters in API_DESIGN.md

### PR DESCRIPTION
**Pull request by LukeWood**

### CODE REVIEW

**Comment by tanzhenyu** [comment]
On a high level this seems a good idea, however there are 2 cases where I'd like to make exceptions:
1) if the number of complex parameters dont exceed 3, given almost every argument is related to some TF ops, but we still pass arguments in most cases. Setting to 3 is a good number, for example let's consider `keras.layers.Dense` -- we don't pass a `kernel_layer` just because there are `kernel_initializer`, `kernel_constraint`, `kernel_regularizer`
2) if we build the original layer to be GPU&TPU compatible, but we cannot make sure whether the passed-in layer argument can be GPU&TPU compatible or not. So while this applies to many conv layers given most tf conv ops work across hardwares, this does not apply to nms layers given most of it don't work across all hardwares.

**Comment by LukeWood** [comment]
> On a high level this seems a good idea, however there are 2 cases where I'd like to make exceptions:
> 
> 1. if the number of complex parameters dont exceed 3, given almost every argument is related to some TF ops, but we still pass arguments in most cases. Setting to 3 is a good number, for example let's consider `keras.layers.Dense` -- we don't pass a `kernel_layer` just because there are `kernel_initializer`, `kernel_constraint`, `kernel_regularizer`
> 2. if we build the original layer to be GPU&TPU compatible, but we cannot make sure whether the passed-in layer argument can be GPU&TPU compatible or not. So while this applies to many conv layers given most tf conv ops work across hardwares, this does not apply to nms layers given most of it don't work across all hardwares.

I don't think the number of arguments should dictate whether or not we expose a wrapper for an op; but rather the consistency of the underlying API.

A good example is NMS: we SHOULD wrap NMS because tf.image makes some strong assertions about formatting.  To achieve API consistency we have to wrap it.  

2.) I could be convinced of... but I'm not fully sure it is worth it at the cost of API consistency.

**Comment by LukeWood** [comment]
> On a high level this seems a good idea, however there are 2 cases where I'd like to make exceptions:
> 
> 1. if the number of complex parameters dont exceed 3, given almost every argument is related to some TF ops, but we still pass arguments in most cases. Setting to 3 is a good number, for example let's consider `keras.layers.Dense` -- we don't pass a `kernel_layer` just because there are `kernel_initializer`, `kernel_constraint`, `kernel_regularizer`
> 2. if we build the original layer to be GPU&TPU compatible, but we cannot make sure whether the passed-in layer argument can be GPU&TPU compatible or not. So while this applies to many conv layers given most tf conv ops work across hardwares, this does not apply to nms layers given most of it don't work across all hardwares.

Another point we should probably consider is that some of our APIs we don't expect many users to use.  For example, AnchorGenerator, ROIGenerator, etc.  These APIs only a small percentage of users will use, so they don't have to be perfect and consistent.  

Top level components, Models, KPLs, etc should in fact be perfect as they are expected to be seen by many many users.

**Comment by tanzhenyu** [comment]
Also,
3. Maybe document how this type of API design works with Keras' config system

**Comment by LukeWood** [comment]
Good point.  I’ll add a section on that and merge

File: .github/API_DESIGN.md [hunk 1]
```diff
@@ -5,11 +5,78 @@ In general, KerasCV abides to the
...
+prefer:
+
+```
+decoder_nms = NonMaxSuppression(
+    threshold=0.05
+    param_two=...,
+    param_three=...
+)
+RetinaNetPredictionDecoder(
+  suppression_layer=decoder_nms
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply .github/API_DESIGN.md:49]: @tanzhenyu maybe this is a way we can ensure GPU compatibility while still abiding to our general design principles?

**Pull request closed without merging**